python/ray/tests/test_cancel.py

```diff
@@ -92,6 +96,179 @@ def dummy(a: SlowToDeserialize):
+
+    check_no_defer_thread = threading.Thread(target=check_no_defer)
+    try:
+        check_no_defer_thread.start()
+        check_no_defer_thread.join()
+    except AssertionError as e:
+        pytest.fail(
+            "DeferSigint.create_if_main_thread() unexpected returned a DeferSigint "
+            f"instance when not in the main thread: {e}"
+        )
+
+    # Check that signal is not deferred when trying to defer it in not the main thread.
+    signal_was_deferred = False
+
+    def maybe_defer():
+        nonlocal signal_was_deferred
+
+        with DeferSigint.create_if_main_thread() as cm:
+            # Check that DeferSigint context manager was NOT returned.
+            assert not isinstance(cm, DeferSigint)
+            # Send singal to current process.
+            os.kill(os.getpid(), signal.SIGINT)
+            # Wait for signal to be delivered.
+            time.sleep(1)
+            # Signal should have been delivered by here, so we consider it deferred if
+            # this is reached.
+            signal_was_deferred = True
+
+    # Create thread that will maybe defer SIGINT.
+    maybe_defer_thread = threading.Thread(target=maybe_defer)
+    try:
+        maybe_defer_thread.start()
+        maybe_defer_thread.join()
+        # KeyboardInterrupt should get raised in main thread.
+    except KeyboardInterrupt:
+        # Check that SIGINT was not deferred.
+        assert not signal_was_deferred
+        # Check that original SIGINT handler was not overridden.
+        assert signal.getsignal(signal.SIGINT) is signal.default_int_handler
+
+
+def test_cancel_during_arg_deser_non_reentrant_import(ray_start_regular):
+    # This test ensures that task argument deserialization properly defers task
+    # cancellation interrupts until after deserialization completes, in order to ensure
+    # that non-reentrant imports that happen during both task argument deserialization
+    # and during error storage are not interrupted.
+
+    # We test this by doing the following:
+    #  - register a custom serializer for (a) a task argument that triggers
+    #  non-reentrant imports on deserialization, and (b) RayTaskError that triggers
+    #  non-reentrant imports on serialization; in our case, we chose pandas it is both
+    #  non-reentrant and expensive, with an import time ~0.5 seconds, giving us a wide
+    #  cancellation target,
+    #  - wait until those serializers are registered on all workers,
+    #  - launch the task and wait until we are confident that the cancellation signal
+    #  will be received by the workers during task argument deserialization (currently a
+    #  200 ms wait).
+    #  - check that a graceful task cancellation error is raised, not a
+    # WorkerCrashedError.
+    def non_reentrant_import():
+        # NOTE: Pandas has a non-reentrant import and should take ~0.5 seconds to
+        # import, giving us a wide cancellation target.
+        import pandas  # noqa
```

Comment:
will it be possible to simulate a super slow `import` here so that when task gets cancelled, it is guaranteed the worker is still executing here, and if so, we could then reliably reproduce an interrupt during import?

Reply:
Easiest method was importing Pandas which is pretty slow; more reliable method would be creating a dummy package that (1) is not reentrant, and (2) sleeps for what we deem to be a sufficient amount of time (e.g. 5 seconds). (1) is the hard part, I still don't know when a package is reentrant and when it's not.

Reply:
why couldn't we `sleep` in `non_reentrant_import` after `import pandas`?

Reply:
@rickyyx The interrupt needs to hit in the middle of the Pandas import, i.e. it needs to interrupt the import; if we sleep after `import pandas` and the interrupt is handled during the sleep, the Pandas import will have completed successfully, and importing Pandas when storing the error will succeed.